### ISSUE TITLE
Block GUI uploads into /systems and /modules paths

### ISSUE THREAD

**Issue by unsoluble** [foundryvtt/foundryvtt#8467]
### User Experience

Currently when you use the Foundry file browser to upload a file into a /systems or /modules path, you get a `warning` banner advising that this is not safe, but the upload happens anyway.

I think that it would be better for file-browser-initiated uploads to these paths to simply be blocked outright, with an `error` banner. Any cases where an upload into these locations legitimately needs to be done for some reason could be done programmatically. Or, if you think there are legitimate use cases that deserve a GUI path, perhaps a manual override could be made available.

**Comment by aaclayton** [comment]
Will keep this open to consider the suggestion, there are valid use cases for this though - for example as content creators we used this feature frequently while working on the creation of A House Divided.

**Comment by unsoluble** [comment]
Yeah, fair point. It would add another layer of dev work, but @MaxPat had the suggestion of a flag you could set in a package's manifest to clear it for uploads — `"uploadAccess": ["true" | "false"]"` or something of the sort. Either that or just a core setting you could throw that opens up the default blockage for everything.

**Comment by Spice-King** [comment]
I'd also tag in that uploading directly into or making folders in `/worlds`, `/systems` and `/modules`  (ex uploading to `/worlds/token.png`, or creating `/system/art/`) should be blocked outright too.

As for uploading into systems and modules, it might as well be a base property in abstract packages system, with default of deny, and worlds overriding it's own default to be allow. This would let GMs (or world package devs, though is there any interest in allowing new worlds on the package listings?) deny uploads to the world's data without blocking all uploads or viewing the world's files. There is also the argument that while running one world, you can't upload into another world's folder.

**Comment by kakaroto** [comment]
I think the API behavior shouldn't change, but the `FileBrowser.canUpload` could simply return `false` for those prefixes, thus removing the upload button and preventing a user's accidental upload to those folders, but without preventing it from the API.
The `developer-mode` module could revert the `canUpload` to return `true` for those folders, thus allowing a developer to upload files to their own module while developing by enabling the option in the developer-mode module. 
I also like the idea of a flag in the module's manifest though I think that might be slightly more complex to use for content creators and there's the risk of forgetting to remove it.

**Comment by unsoluble** [comment]
> I'd also tag in that uploading directly into or making folders in `/worlds` ... should be blocked outright too.

Storing user data in `/worlds` is an extremely common (and oft recommended) usage pattern, especially when portability is desired.

**Comment by kakaroto** [comment]
> > I'd also tag in that uploading directly into or making folders in `/worlds` ... should be blocked outright too.
> 
> Storing user data in `/worlds` is an extremely common (and oft recommended) usage pattern, especially when portability is desired.

Yeah, the world shouldn't be blocked, since you can't "accidentally update or uninstall" a world. It's expected that if you do, you'd be losing data

**Comment by aaclayton** [comment]
Raising this for consideration in V11 since we have added the persistent storage location which interacts with this suggestion.

**Comment by mhilbrunner** [comment]
I agree with kakaroto's suggestions. I agree it makes sense to make it a little harder to do things that are ill-advised if you don't know what you're doing and that may cause data loss.

At the same time, Foundry's openness and how easy it is to tinker on it is one of its strength to me (and like with Linux, that openness _does_ of course come with some footguns, great power and all that).

So IMO its fine to disallow these, but at the same time I'd consider a "I know what I'm doing and take responsibility if I lose data" toggle (if that's a hidden property that can be set via a "dev mode" module, thats fine by me).

**Comment by Spice-King** [comment]
> > I'd also tag in that uploading directly into or making folders in `/worlds` ... should be blocked outright too.
> 
> Storing user data in `/worlds` is an extremely common (and oft recommended) usage pattern, especially when portability is desired.

Oh yea, should have been more clear, inside an active world's folder was noted below. Just wanted to avoid uploading to the root of the worlds folder and other worlds that's not the currently running world.

**Comment by mattexdee** [comment]
Added to the V11 milestone so we can consider options for uploading to system and/or module folders to support systems that provide unique icons based on Actor type.

**Comment by caewok** [comment]
Note that modules should have a way to override this. Elevated Vision, for example, stores terrain elevation maps in the world folder, identified by scene id. This is so the module can be reinstalled without losing the terrain maps for different scenes, some of which may have been complicated to set up. The user does not initiate this save, but EV uses the same underlying methods to save. So a toggle of some kind could be used to allow saves when it is intentional.

**Comment by unsoluble** [comment]
> Note that modules should have a way to override this.

To be clear, the request here is _specifically_ for GUI uploading to be prevented.

**Comment by MaxPat931** [comment]
I do think there should still be a myhod for users to be able to use the GUI, I do this frequently for my own personal content sharing modules. It's much easier to add one or two assets for a new npc for instance rather than having to switch programs, to remote into my server and drop a file in.

**Comment by Fyorl** [comment]
Prevents the following uploads via GUI only:
- Prevents uploading to the root of `worlds/`, `systems/`, or `modules/`.
- If a World is active, prevents uploading to a World directory or System directory that is not the active one.
- Prevents uploading to any Module or System directory, unless that package has the `canUpload` flag in its metadata.

**Comment by JPMeehan** [comment]
As part of this, will the KB be updated to list all core-supported flags? Trying to help new devs find the docs for stuff like hotReload is unusually onerous.
